### PR TITLE
#1176 - Support function for Rectification

### PR DESCRIPTION
**Pull request by schillic**
See #1176.

### CODE REVIEW

**Comment by schillic** [comment]
In principle this works. Compare the [example here](https://github.com/JuliaReach/LazySets.jl/issues/1176#issuecomment-490396060) with this result:
rectification
Here I plotted the four intersections/projections (the origin, two line segments, and the polygon) separately.

There are some precision problems due to lazy intersections. For instance, `ρ([-1., -1.], RP) ~ 0.1`, but it should clearly be 0. If I perform a concrete intersection instead, everything works.

File: src/Rectification.jl [hunk 1]
```diff
@@ -353,3 +411,131 @@ function isbounded(r::Rectification{N})::Bool where {N<:Real}
...
+        for (i, d) in enumerate(mixed_dimensions)
+            if mixed_dimensions_enumeration[i]
+                # reset mixed index to one again because the set is nonnegative
+                v[d] = one(N)
+            end
+        end
+        return Diagonal(v) * set
+    end
+
+    # identify dimensions where set is negative and mixed (positive/negative)
```
Comment (mforets) [inline src/Rectification.jl]: Let's outsource the algorithm that receives a set and returns the indices where the set is negative (or is positive/negative) algorithm to a separate function.

File: src/Rectification.jl [hunk 2]
```diff
@@ -353,3 +411,131 @@ function isbounded(r::Rectification{N})::Bool where {N<:Real}
...
+    negative_dimensions = Vector{Int}()
+    mixed_dimensions = Vector{Int}()
+    mixed_dimensions_enumeration = BitArray(undef, 0)
+    for i in 1:n
+        d = Approximations.UnitVector(i, n, -one(N))
```
Comment (mforets) [inline src/Rectification.jl]: i'd prefer to see `d = UnitVector(i, n, -one(N))` let's `using Approximations: UnitVector` if needed on top of this file

File: src/Rectification.jl [hunk 3]
```diff
@@ -353,3 +411,131 @@ function isbounded(r::Rectification{N})::Bool where {N<:Real}
...
+        if ρ(d, X) > zero(N)
+            push!(negative_dimensions, i)
+            d = Approximations.UnitVector(i, n, one(N))
+            if ρ(d, X) > zero(N)
```
Comment (mforets) [inline src/Rectification.jl]: we can do `ρ(-d, X) > zero(N)` which is efficient and also returns a `SingleEntryVector` (https://github.com/JuliaReach/LazySets.jl/pull/1523)

File: src/Rectification.jl [hunk 4]
```diff
@@ -353,3 +411,131 @@ function isbounded(r::Rectification{N})::Bool where {N<:Real}
...
+                push!(mixed_dimensions, i)
+            end
+        end
+    end
+
+    if isempty(mixed_dimensions)
+        # no mixed dimensions: no intersections needed
+        if isempty(negative_dimensions)
+            # no negative dimensions: set is unaffected by rectification
+            r.cache.set = X
```
Comment (mforets) [inline src/Rectification.jl]: The set `X` may be unaffected by the rectification. So if we change for `convert` we have to wrap it around `UnionSetArray`..

File: src/Rectification.jl [hunk 5]
```diff
@@ -353,3 +411,131 @@ function isbounded(r::Rectification{N})::Bool where {N<:Real}
...
+            # project negative dimensions to zero
+            v = ones(N, n)
+            v[negative_dimensions] .= zero(N)
+            r.cache.set = Diagonal(v) * X
+        end
+    else
+        # apply intersections with half-spaces in every mixed dimension
+        m = length(mixed_dimensions)
+        i = m
+        projections = Vector{LazySet{N}}()
```
Comment (mforets) [inline src/Rectification.jl]: Aren't these projections instances of `LinearMap`?

File: src/Rectification.jl [hunk 6]
```diff
@@ -353,3 +411,131 @@ function isbounded(r::Rectification{N})::Bool where {N<:Real}
...
+        sizehint!(projections, 2^m)
+        while true
+            # compute next projection
+            polyhedron = construct_constraints(mixed_dimensions,
+                                               mixed_dimensions_enumeration)
+            projection = construct_projection(X ∩ polyhedron,
```
Comment (mforets) [inline src/Rectification.jl]: if it's not too hard maybe add an option to compute this intersection either concretely or lazily

File: src/Rectification.jl (hunk 4 above)
Comment (schillic) [reply src/Rectification.jl]: True, so I will not use `convert`.

**Pull request merged**